### DISCUSSION TITLE
How to idiomatically extract all fluents present in a formula?

### DISCUSSION THREAD

**Discussion by JoanEspasa** [aiplan4eu/unified-planning#600] [Q&A] [answered]
I'm doing some preprocessing of the preconditions of a problem, and I need to extract all variables present in them. Once I get a grounded action and access its precondition, I have been able to extract its variables by using: `pre.environment.free_vars_extractor.get(pre)` 

It feels a bit strange having to access the environment of the precondition to have access to the walker for then calling the `get` method. Is there a more idiomatic/clean way of getting these?

Thanks and apologies in advance if there is a more obvious way that I've overlooked!

**Comment by Framba-Luca** [comment, marked as answer]
Hi @JoanEspasa , for the `simplification` and `substitution` we made a "shortcut" (like pre.simplify()), but not for the extraction of variables; this might be a useful shortcut to add even if I am not entirely sure due to the complex and not immediate naming of that functionality. 

Currently, the cleanest way to do it is to store the `free_vars_extractor` of the `environment` in a variable and call it.

For example:
```python
fve = problem.environment.free_vars_extractor
...
pre_free_variables = fve.get(pre)
```

**Reply by JoanEspasa** [reply]
Thanks for clarifying this Framba, it is appreciated :)
